crates/miden-objects/src/block/proven_block.rs

```diff
@@ -0,0 +1,156 @@
+///   and note metadata while for public notes the full note details are included.
+/// - A list of new nullifiers created for all notes that were consumed in the block.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ProvenBlock {
+    /// The header of the block, committing to the current state of the chain.
+    header: BlockHeader,
+
+    /// Account updates for the block.
+    updated_accounts: Vec<BlockAccountUpdate>,
+
+    /// Note batches created by the transactions in this block.
+    output_note_batches: Vec<OutputNoteBatch>,
+
+    /// Nullifiers created by the transactions in this block through the consumption of notes.
+    created_nullifiers: Vec<Nullifier>,
+}
+
+impl ProvenBlock {
+    /// Returns a new [`ProvenBlock`] instantiated from the provided components.
+    ///
+    /// # Warning
+    ///
+    /// This constructor does not do any validation, so passing incorrect values may lead to later
+    /// panics.
+    pub fn new_unchecked(
+        header: BlockHeader,
+        updated_accounts: Vec<BlockAccountUpdate>,
+        output_note_batches: Vec<OutputNoteBatch>,
+        created_nullifiers: Vec<Nullifier>,
+    ) -> Self {
+        Self {
+            header,
+            updated_accounts,
+            output_note_batches,
+            created_nullifiers,
+        }
+    }
+
+    /// Returns a commitment to this block.
+    pub fn hash(&self) -> Digest {
+        self.header.hash()
+    }
+
+    /// Returns the header of this block.
+    pub fn header(&self) -> BlockHeader {
+        self.header
+    }
+
+    /// Returns the slice of [`BlockAccountUpdate`]s for all accounts updated in this block.
+    pub fn updated_accounts(&self) -> &[BlockAccountUpdate] {
+        &self.updated_accounts
+    }
+
+    /// Returns the slice of [`OutputNoteBatch`]es for all output notes created in this block.
+    pub fn output_note_batches(&self) -> &[OutputNoteBatch] {
+        &self.output_note_batches
+    }
+
+    /// Returns an iterator over all notes created in this block.
+    ///
+    /// Each note is accompanied by a corresponding index specifying where the note is located
+    /// in the block's note tree.
+    pub fn notes(&self) -> impl Iterator<Item = (BlockNoteIndex, &OutputNote)> {
+        self.output_note_batches.iter().enumerate().flat_map(|(batch_idx, notes)| {
+            notes.iter().enumerate().map(move |(note_idx_in_batch, note)| {
+                (
+                    // SAFETY: The proven block contains at most the max allowed number of batches
+                    // and each batch is guaranteed to contain at most the
+                    // max allowed number of output notes.
+                    BlockNoteIndex::new(batch_idx, note_idx_in_batch),
+                    note,
+                )
+            })
+        })
+    }
+
+    /// Returns the [`BlockNoteTree`] containing all notes created in this block.
+    pub fn build_note_tree(&self) -> BlockNoteTree {
```

Comment:
nit: should we rename these into `output_notes()` and `build_output_note_tree()` for consistency?